lib/location/method_gnss.c

```diff
@@ -0,0 +1,791 @@
+static void method_gnss_manage_pgps(struct k_work *work)
+{
+	ARG_UNUSED(work);
+	int err;
+
+	LOG_DBG("Sending prediction to modem...");
+
+	err = nrf_cloud_pgps_inject(prediction, &agps_request);
+	if (err) {
+		LOG_ERR("Unable to send prediction to modem: %d", err);
+	}
+
+	err = nrf_cloud_pgps_preemptive_updates();
+	if (err) {
+		LOG_ERR("Error requesting updates: %d", err);
+	}
+}
+
+void method_gnss_pgps_handler(struct nrf_cloud_pgps_event *event)
+{
+	LOG_DBG("P-GPS event type: %d", event->type);
+
+	if (event->type == PGPS_EVT_AVAILABLE) {
+		prediction = event->prediction;
+		k_work_submit_to_queue(location_core_work_queue_get(),
+				       &method_gnss_manage_pgps_work);
+	} else if (event->type == PGPS_EVT_REQUEST) {
+		memcpy(&pgps_request, event->request, sizeof(pgps_request));
+		k_work_submit_to_queue(location_core_work_queue_get(),
+				       &method_gnss_pgps_request_work);
+	}
+}
+
+static void method_gnss_notify_pgps(struct k_work *work)
+{
+	ARG_UNUSED(work);
+	int err = nrf_cloud_pgps_notify_prediction();
+
+	if (err) {
+		LOG_ERR("Error requesting notification of prediction availability: %d", err);
+	}
+}
+#endif
+
+void method_gnss_lte_ind_handler(const struct lte_lc_evt *const evt)
+{
+	switch (evt->type) {
+	case LTE_LC_EVT_MODEM_SLEEP_ENTER:
+		if (evt->modem_sleep.type == LTE_LC_MODEM_SLEEP_PSM) {
+			/* Allow GNSS operation once LTE modem is in power saving mode. */
+			k_sem_give(&entered_psm_mode);
+		}
+		break;
+	case LTE_LC_EVT_MODEM_SLEEP_EXIT:
+		/* Prevent GNSS from starting while LTE is active. */
+		k_sem_reset(&entered_psm_mode);
+		break;
+	case LTE_LC_EVT_PSM_UPDATE:
+		/* If PSM becomes disabled e.g. due to network change, allow GNSS to be started
+		 * in case there was a pending position request waiting for the sleep to start. If
+		 * PSM becomes enabled, block GNSS until the modem enters PSM by taking the
+		 * semaphore.
+		 */
+		if (evt->psm_cfg.active_time == -1) {
+			k_sem_give(&entered_psm_mode);
+		} else if (evt->psm_cfg.active_time > 0) {
+			k_sem_take(&entered_psm_mode, K_NO_WAIT);
+		}
+		break;
+	case LTE_LC_EVT_RRC_UPDATE:
+		if (evt->rrc_mode == LTE_LC_RRC_MODE_CONNECTED) {
+			/* Prevent GNSS from starting while RRC is in connected mode. */
+			k_sem_reset(&entered_rrc_idle);
+		} else if (evt->rrc_mode == LTE_LC_RRC_MODE_IDLE) {
+			/* Allow GNSS operation once RRC is in idle mode. */
+			k_sem_give(&entered_rrc_idle);
+		}
+		break;
+	default:
+		break;
+	}
+}
+
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+#if defined(CONFIG_NRF_CLOUD_MQTT)
+static void method_gnss_agps_request_work_fn(struct k_work *item)
+{
+	int err = nrf_cloud_agps_request(&agps_request);
+
+	if (err) {
+		LOG_ERR("nRF Cloud A-GPS request failed, error: %d", err);
+		return;
+	}
+
+	LOG_DBG("A-GPS data requested");
+}
+
+#else /* defined(CONFIG_NRF_CLOUD_MQTT) */
+static int method_gnss_get_modem_info(struct lte_lc_cell *serving_cell)
+{
+	__ASSERT_NO_MSG(serving_cell != NULL);
+
+	int err = modem_info_init();
+
```

Comment:
Why not use `location_utils_modem_params_read()` rather than duplicate the functionality here?

Reply:
Because it did not exist when I wrote this code :) I'll make use of this function

Reply:
This is now changed